### PR TITLE
ENT-3476: cf-testd, a mock version of cf-serverd

### PR DESCRIPTION
**Pull request by olehermanse**
Must be merged with the enterprise functions:
https://github.com/cfengine/enterprise/pull/374

### CODE REVIEW

**Comment by olehermanse** [comment]
@jimis @vpodzime Please review.

File: cf-testd/Makefile.am [hunk 1]
```diff
@@ -0,0 +1,57 @@
+#
+#  Copyright 2017 Northern.tech AS
```
Comment (jimis) [inline cf-testd/Makefile.am]: year

File: cf-testd/cf-testd.c [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+    "Log timestamps on each line of log output",
+    "Output verbose information about the behaviour of the agent",
+    "Output the version of the software",
+    NULL
+};
+
+CFTestD_Config *CFTestD_ConfigInit()
+{
+    CFTestD_Config* r = (CFTestD_Config *)(xmalloc(sizeof(CFTestD_Config)));
+    r->file = NULL;
```
Comment (vpodzime) [inline cf-testd/cf-testd.c]: ``xcalloc()`` can be used here making the ``r->file`` assignment unnecessary.

File: cf-testd/cf-testd.c [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+static void CFTestD_DeleteConn(ServerConnectionState *conn)
+{
+    int sd = ConnectionInfoSocket(conn->conn_info);
+    if (sd != SOCKET_INVALID)
+    {
+        cf_closesocket(sd);
+    }
+    ConnectionInfoDestroy(&conn->conn_info);
+
+    *conn = (ServerConnectionState) {0};
```
Comment (vpodzime) [inline cf-testd/cf-testd.c]: What is this line about?

File: cf-testd/cf-testd.c [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+    strlcpy(aligned_ipaddr, conn->ipaddr, sizeof(aligned_ipaddr));
+    strlcat(aligned_ipaddr, "> ",         sizeof(aligned_ipaddr));
+    /* Pad with enough spaces for IPv4 addresses to be aligned. Max chars are
+     * 15 for the address plus two for "> " == 17. */
+    size_t len;
+    for (len = strlen(aligned_ipaddr); len < 17; len++)
+    {
+        aligned_ipaddr[len] = ' ';
+    }
+    aligned_ipaddr[len] = '\0';
```
Comment (vpodzime) [inline cf-testd/cf-testd.c]: This doesn't seem to be used anywhere.

File: cf-testd/cf-testd.c [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+    bool established = CFTestD_TLSSessionEstablish(conn);
+    if (!established)
+    {
+        Log(LOG_LEVEL_ERR, "Could not establish TLS Session");
+        return NULL;
+    }
+    int ret = getnameinfo((const struct sockaddr *) &conn->conn_info->ss,
+                      conn->conn_info->ss_len,
+                      conn->revdns, sizeof(conn->revdns),
+                      NULL, 0, NI_NAMEREQD);
```
Comment (vpodzime) [inline cf-testd/cf-testd.c]: A bit weird alignment.

File: cf-testd/cf-testd.c [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+    /* ============================================================ */
+
+    CFTestD_DeleteConn(conn);
+    return NULL;
+}
+
+static void CFTestD_SpawnConnection(const char *ipaddr, ConnectionInfo *info)
+{
+    ServerConnectionState *conn = NULL;
+    conn = CFTestD_NewConn(info);
```
Comment (vpodzime) [inline cf-testd/cf-testd.c]: I think these two lines can be squashed. Or is the combination way too long?

File: cf-testd/cf-testd.c [hunk 7]
```diff
@@ -0,0 +1,492 @@
...
+    ConnectionInfoSocket(info);
+    strlcpy(conn->ipaddr, ipaddr, CF_MAX_IP_LEN );
```
Comment (vpodzime) [inline cf-testd/cf-testd.c]: Extra space after ``CF_MAX_IP_LEN``

File: cf-testd/cf-testd.c [hunk 8]
```diff
@@ -0,0 +1,492 @@
...
+
+    Log(LOG_LEVEL_WARNING, "Thread is being handled from main loop!");
```
Comment (vpodzime) [inline cf-testd/cf-testd.c]: *Thread is being handled...* ?

File: cf-testd/cf-testd.c [hunk 9]
```diff
@@ -0,0 +1,492 @@
...
+    SendTransaction(conn->conn_info, sendbuffer, 0, CF_DONE);
+    Log(LOG_LEVEL_INFO,
+        "Closing connection due to illegal request: %s", recvbuffer);
+    return false;
+}
+
+static bool CFTestD_BusyLoop(ServerConnectionState *conn)
+{
+    char recvbuffer[CF_BUFSIZE + CF_BUFEXT] = { 0 };
+    char sendbuffer[CF_BUFSIZE - CF_INBAND_OFFSET] = { 0 };
```
Comment (vpodzime) [inline cf-testd/cf-testd.c]: Other empty string initializers are ``""``, maybe these should use the same syntax?

File: cf-testd/cf-testd.c [hunk 10]
```diff
@@ -0,0 +1,492 @@
...
+{
+    bool tls_init_ok = ServerTLSInitialize();
+    if (!tls_init_ok)
+    {
+        return -1;
+    }
+
+    int sd = InitServer(10);
+
+    while (true)
```
Comment (vpodzime) [inline cf-testd/cf-testd.c]: I'd prefer ``while (selected != -1)`` here, but YMMV.

File: cf-testd/cf-testd.c [hunk 11]
```diff
@@ -0,0 +1,492 @@
...
+
+int main(int argc, char *argv[])
+{
+    Log(LOG_LEVEL_VERBOSE, "Starting cf-testd");
+    CryptoInitialize();
+    LoadSecretKeys();
+    cfnet_init(NULL, NULL);
+    CFTestD_Config *config = CFTestD_CheckOpts(argc, argv);
+
+    if(config->file != NULL)
```
Comment (vpodzime) [inline cf-testd/cf-testd.c]: Missing space after ``if``

File: cf-testd/cf-testd.c [hunk 12]
```diff
@@ -0,0 +1,492 @@
...
+static const char *const CF_TESTD_MANPAGE_LONG_DESCRIPTION =
+        "Used for testing CFEngine by mimicing cf-serverd.";
+
+static const struct option OPTIONS[] =
+{
+    {"debug",     no_argument,       0, 'd'},
+    {"file",      required_argument, 0, 'f'},
+    {"help",      no_argument,       0, 'h'},
+    {"inform",    no_argument,       0, 'I'},
+    {"timestamp", no_argument,       0, 'l'},
```
[the remark this answers is not in the document]
Comment (olehermanse) [reply cf-testd/cf-testd.c]: Because it's the same in cf-serverd.

File: cf-testd/cf-testd.c (hunk 8 above)
Comment (olehermanse) [reply cf-testd/cf-testd.c]: This is also from cf-serverd ;) Should be "Connection is being handled from main loop!" imo

File: cf-testd/Makefile.am (hunk 1 above)
Comment (olehermanse) [reply cf-testd/Makefile.am]: Fixed.

[the remark this answers is not in the document]
Comment (olehermanse) [reply cf-testd/Makefile.am]: Fixed.

File: cf-testd/cf-testd.c [hunk 13]
```diff
@@ -0,0 +1,492 @@
...
+}CFTestD_Config;
+
+/*******************************************************************/
+/* Command line option parsing                                     */
+/*******************************************************************/
+
+static const char *const CF_TESTD_SHORT_DESCRIPTION = "CFEngine dummy server";
+
+static const char *const CF_TESTD_MANPAGE_LONG_DESCRIPTION =
+        "Used for testing CFEngine by mimicing cf-serverd.";
```
[the remark this answers is not in the document]
Comment (olehermanse) [reply cf-testd/cf-testd.c]: They were for the man page of cf-serverd. Don't think cf-testd needs one. Removed.

File: cf-testd/cf-testd.c [hunk 14]
```diff
@@ -0,0 +1,492 @@
...
+    }
+
+    return config;
+}
+
+bool CFTestD_TLSSessionEstablish(ServerConnectionState *conn)
+{
+    if (conn->conn_info->status == CONNECTIONINFO_STATUS_ESTABLISHED)
+    {
+        return 1;
```
[the remark this answers is not in the document]
Comment (olehermanse) [reply cf-testd/cf-testd.c]: Thanks, fixed.

File: cf-testd/cf-testd.c [hunk 15]
```diff
@@ -0,0 +1,492 @@
...
+    ServerConnectionState *conn, char *recvbuffer, int encrypt)
+{
+    char query[CF_BUFSIZE];
+
+    query[0] = '\0';
+    sscanf(recvbuffer, "QUERY %255[^\n]", query);
+
+    if (strlen(query) == 0)
+    {
+        return false;
```
[the remark this answers is not in the document]
Comment (olehermanse) [reply cf-testd/cf-testd.c]: Thanks, fixed.

File: cf-testd/cf-testd.c (hunk 2 above)
Comment (olehermanse) [reply cf-testd/cf-testd.c]: Thanks, fixed.

File: cf-testd/cf-testd.c (hunk 3 above)
Comment (olehermanse) [reply cf-testd/cf-testd.c]: Don't know. Copied from cf-serverd. Maybe @jimis can elaborate?

File: cf-testd/cf-testd.c (hunk 4 above)
Comment (olehermanse) [reply cf-testd/cf-testd.c]: It is used in the log context.

File: cf-testd/cf-testd.c (hunk 5 above)
Comment (olehermanse) [reply cf-testd/cf-testd.c]: Thanks, fixed.

File: cf-testd/cf-testd.c (hunk 6 above)
Comment (olehermanse) [reply cf-testd/cf-testd.c]: Thanks, fixed.

File: cf-testd/cf-testd.c (hunk 7 above)
Comment (olehermanse) [reply cf-testd/cf-testd.c]: Thanks.

File: cf-testd/cf-testd.c (hunk 8 above)
Comment (olehermanse) [reply cf-testd/cf-testd.c]: Fixed in cf-testd. Will remember to change message in serverd as well.

File: cf-testd/cf-testd.c (hunk 9 above)
Comment (olehermanse) [reply cf-testd/cf-testd.c]: Agree, fixed.

File: cf-testd/cf-testd.c (hunk 10 above)
Comment (olehermanse) [reply cf-testd/cf-testd.c]: Fixed.

File: cf-testd/cf-testd.c (hunk 11 above)
Comment (olehermanse) [reply cf-testd/cf-testd.c]: Fixed.

File: cf-testd/cf-testd.c (hunk 4 above)
Comment (olehermanse) [reply cf-testd/cf-testd.c]: Or, that is what it is used for in cf-serverd. Removing it.

**Comment by olehermanse** [comment]
https://ci.cfengine.com/view/all/job/pr-pipeline/81/

File: cf-testd/cf-testd.c (hunk 3 above)
Comment (jimis) [reply cf-testd/cf-testd.c]: zeroing the memory?

Comment (vpodzime) [reply cf-testd/cf-testd.c]: Could be, but the zeroed memory is used (the pointer is dereferenced again) on the line below this.

**Comment by olehermanse** [comment]
Completely green. Merging.

**Pull request merged**